Make backporting responsibility more clear

joestringer opened the pull request:
There have been a few occasions recently where backporters have resolved
conflicts but did not notify the authors or raise any concern about
the conflicts. This can be potentially problematic, because close review
is primarily performed when merging the initial PR and it is easy to pay
less attention to backport PRs. If backport commits are not prepared
carefully, we can introduce regressions in the backport version of the
change when the upstream version of the change is correct.

In order to hopefully reduce the chance of such an issue, clarify that
the backporter has the responsibility to review the changes they are
proposing before they submit the PR. If we make the mechanical steps
clear to the backporter, then they should be able to take these
precautions to either confirm that the changes really are the same, or
at the very least become aware of which areas differ and use that to
raise a discussion with the original author of the PR being backported.
The review process then can be more targeted around having the original
author ensure the backport has the same behaviour as the original PR,
given the different code in the target branch.

joestringer committed:
docs: Make backporting responsibility more clear

There have been a few occasions recently where backporters have resolved
conflicts but failed to notify the authors or raise any concern about
the conflicts. This can be potentially problematic, because close review
is primarily performed when merging the initial PR and it is easy to pay
less attention to backport PRs. If backport commits are not prepared
carefully, we can introduce regressions in the backport version of the
change when the upstream version of the change is correct.

In order to hopefully reduce the chance of such an issue, clarify that
the backporter has the responsibility to review the changes they are
proposing before they submit the PR. If we make the mechanical steps
clear to the backporter, then they should be able to take these
precautions to either confirm that the changes really are the same, or
at the very least become aware of which areas differ and use that to
raise a discussion with the original author of the PR being backported.
The review process then can be more targeted around having the original
author ensure the backport has the same behaviour as the original PR.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on Documentation/contributing/release/backports.rst:
```diff
@@ -178,6 +178,15 @@ Creating the Backports Branch
    backport reviewers towards which changes may deviate from the original
    commits to ensure that the changes are correctly backported.
 
```
We should also document that backporters should look for current outstanding backports (https://github.com/cilium/cilium/pulls?q=is%3Aopen+is%3Apr+label%3Akind%2Fbackports) and reach out to the previous backporter (either DM or in #launchpad) to discuss those backports and how they can help get them in.

aanm reviewed (approved):
LGTM, including the comment left

christarazi reviewed (approved):
I think this is a good change. 

One thing that I would like to see more widespread use of is providing context of how the conflict was resolved in the backported commit. This practice I've picked up from Joe which is to have a `[ Backporter's notes: ... ]` in the commit msg where this information is shared. In general, having more context on what the conflict was and how it was resolved can be impactful when debugging regressions in older versions.

qmonnet reviewed (approved):
Looks good to me.

```
$ vim -O2 -c "windo set scrollbind" <(git show) \
        <(git show $(git show | awk '/\[ upstream commit / {print $4}'))
```

joestringer committed:
docs: Document to minimize outstanding backport PRs

Having multiple outstanding backport PRs makes it more likely that there
can be conflicts between the two, and diverges from the goal of ensuring
that all intended backports are merged into the tree.

Encourage backporters to first reach out to authors of existing backport
PRs to help coordinate how to get the existing PRs triaged, reviewed and
merged before creating new backport PRs.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
docs: Document backporter's notes commit example

Document an example of how to describe how conflicts were resolved.

Signed-off-by: Joe Stringer <joe@cilium.io>
